ethsigner/core/src/main/java/tech/pegasys/ethsigner/core/signing/hashicorp/HashicorpSignerHelper.java

```diff
@@ -0,0 +1,138 @@
+    if (tokenString == null) {
+      return null;
+    }
+    final String response = requestSecretFromVault(config, tokenString);
+    if (response == null) {
+      return null;
+    }
+    final Credentials credentials = extractCredentialsFromJson(response);
+    if (credentials == null) {
+      return null;
+    }
+    return new CredentialTransactionSigner(credentials);
+  }
+
+  private static String readTokenFromFile(HashicorpSignerConfig config) {
+    List<String> authFileLines;
+    try {
+      authFileLines = Files.readAllLines(config.getAuthFilePath());
+    } catch (IOException e) {
+      LOG.error(
+          "Unable to read file containing the authentication information for Hashicorp Vault.\n"
+              + config.getAuthFilePath()
+              + "\n"
+              + e.getMessage());
+      return null;
+    }
+    return authFileLines.get(0);
+  }
+
+  private static String requestSecretFromVault(HashicorpSignerConfig config, String tokenString) {
+    final String requestURI = "/v1" + config.getSigningKeyPath();
+
+    return getVaultResponse(config, tokenString, requestURI);
+  }
+
+  private static String getVaultResponse(
+      HashicorpSignerConfig config, String tokenString, String requestURI) {
+    final CompletableFuture<String> future = new CompletableFuture<>();
+    final HttpClientRequest request =
+        hashicorpVaultClient.request(
+            HttpMethod.GET,
+            config.getServerPort(),
+            config.getServerHost(),
+            requestURI,
+            rh -> {
+              rh.bodyHandler(
+                  bh -> {
+                    if (rh.statusCode() == 200) {
+                      future.complete(bh.toString());
+                    } else {
+                      future.completeExceptionally(
+                          new Exception(
+                              "Hashicorp vault responded with status code {"
+                                  + rh.statusCode()
+                                  + "}"));
+                    }
+                  });
+            });
+    if (tokenString != null) {
+      request.headers().set("X-Vault-Token", tokenString);
+    }
+    request.setChunked(false);
+    request.end();
+    String response;
+    try {
+      response = future.get(config.getTimeout(), TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException e) {
+      LOG.error(
+          "Unable to retrieve private key from Hashicorp Vault.\n"
+              + config.toString()
+              + "\n"
+              + e.getMessage());
+      return null;
+    } catch (TimeoutException e) {
+      LOG.error(
+          "Timeout while retrieving private key from Hashicorp Vault.\n"
+              + config.toString()
+              + "\n"
+              + e.getMessage());
+      return null;
+    }
+    return response;
+  }
+
+  private static Credentials extractCredentialsFromJson(String response) {
+    if (response == null) {
```

Comment:
need to think about how we handle the continual passing through nulls  .... this works but potentially hides when a problem occurs (or at least delays it).

Reply:
I'm not sure about that either. I've just copied the way it was done before. Maybe we should have a ticket to change to fail early where it makes sense.